### ISSUE TITLE
Conjugate_Gradient Tolerance

### ISSUE THREAD

**Issue by Zeppo1994** [deepinv/deepinv#992]
### Environment

DeepInverse Main Branch

### Describe the bug

When getting passed a zero vector b, the conjugate_gradient algorithm sets the tolerance to zero and iterates until it hits maxiter:

 tol = dot(b, b, dim=dim).real * (tol**2)

While probably no-one would pass a zero vector intentionally, it can  happen in the backward pass of implicit differentiation techniques. I think we should put a safeguard and check if this also applies to the other linear solvers.

### Steps to reproduce

None.

### Expected behavior

The tolerance should never be set to zero.

**Comment by jscanvic** [comment]
Thank you for raising this issue! Any suggestion for a better value of the tolerance in that case?

**Comment by Zeppo1994** [comment]
If we do not want to break anything by accident, we can simply do a check if the rhs b is zero. Then, we can break the iteration and return an all zero tensor instead.

Not sure if the complex case is tested. If so, it would also make more sense to use r.norm(dim=dim) ** 2 as rescaling factor for the problem.

Suggestions (I hope that I did not miss anything):
b_norm = r.norm(dim=dim) ** 2 
if b_norm == 0:
    if verbose:
        print("CG Converged at iteration 0")
    return zeros_like(b)
tol = b_norm * (tol**2)

**Comment by jscanvic** [comment]
Ah alright, thank you! We'll take care of it shortly!

**Comment by Zeppo1994** [comment]
Since, we are already on it: Is there a good reason for making the tolerance adaptive to \vert b \vert whereas eps is not adapted? For several problems, I observed that I have to decrease the eps in order to ensure that CG converges.

**Comment by Zeppo1994** [comment]
Here is a full modified version of the original code, removing the eps and dealing with the potentially zero res b. We can discuss if an epsilon should be actually included or not. Typically, we should only see problems if A is (almost) indefinite and in that case it would be better to resort to minres anyways. The masking mechanism now deals with the fact that we could have divisions by zero since some of the systems in the batch are already converged.

def conjugate_gradient(
    A: Callable,
    b: Tensor,
    max_iter: int = 1e2,
    tol: float = 1e-5,
    parallel_dim: None | int | list[int] = 0,
    init: Tensor = None,
    verbose: bool = False,
) -> Tensor:
    """
    Standard conjugate gradient algorithm for batched computations.
    
    It solves the linear system :math:`Ax=b`, where :math:`A` is a (square) linear operator and :math:`b` is a tensor.
    
    For more details see: http://en.wikipedia.org/wiki/Conjugate_gradient_method
    
    :param Callable A: Linear operator as a callable function, has to be square!
    :param torch.Tensor b: input tensor of shape (B, ...)
    :param int max_iter: maximum number of CG iterations
    :param float tol: absolute tolerance for stopping the CG algorithm.
    :param None, int, list[int] parallel_dim: dimensions to be considered as batch dimensions. If None, all dimensions are considered as batch dimensions.
    :param torch.Tensor init: Optional initial guess.
    :param bool verbose: Output progress information in the console.
     :return: torch.Tensor :math:`x` of shape (B, ...) verifying :math:`Ax=b`.
    """
    if isinstance(parallel_dim, int):
        parallel_dim = [parallel_dim]
    if parallel_dim is None:
        parallel_dim = []
    if isinstance(b, TensorList):
        dim = [i for i in range(b[0].ndim) if i not in parallel_dim]
    else:
        dim = [i for i in range(b.ndim) if i not in parallel_dim]
    if init is not None:
        x = init
    else:
        x = torch.zeros_like(b)
    r = b - A(x)
    p = r
    rsold = dot(r, r, dim=dim).real
    b_norm_sq = dot(b, b, dim=dim).real
    b_norm_sq = torch.where(b_norm_sq > 0, b_norm_sq, torch.ones_like(b_norm_sq))
    tol_sq = b_norm_sq * tol**2
    active = rsold >= tol_sq
    
    for iteration in range(int(max_iter)):
        if not torch.any(active):
            if verbose:
                print("CG Converged at iteration", iteration)
            break
        
        Ap = A(p)
        pAp = dot(p, Ap, dim=dim).real
        
        alpha = torch.where(active, rsold / pAp, torch.zeros_like(rsold))
        
        x = x + p * alpha
        r = r - Ap * alpha
        rsnew = dot(r, r, dim=dim).real
        
        beta = rsnew / rsold
        
        active = rsnew >= tol_sq
        beta = torch.where(active, beta, torch.zeros_like(beta))
        
        p = r + p * beta
        rsold = rsnew
        
        if iteration > 0 and iteration % 50 == 0 and torch.any(active):
            r_true = b - A(x)
            r = torch.where(active.unsqueeze(-1).expand_as(r), r_true, r)
            rsold = dot(r, r, dim=dim).real
    else:
        if verbose:
            print("CG did not converge")
    
    return x
